File: prototype/java/Geometry.java
abstract class Shape implements Cloneable {

	private int x, y;
	private String color;

	Shape() {
		this(140, 210, "Red");
	}

	Shape(int x, int y, String color) {
		this.x = x;
		this.y = y;
		this.color = color;
	}

	protected Shape(Cloneable prototype) {
		Shape shape = (Shape)prototype;
		this.x = shape.x;
		this.y = shape.y;
		this.color = shape.color;
	}

	@Override
	public boolean equals(Object another) {
		if (another == null || !(another instanceof Shape)) return false;

		Shape shape = (Shape)another;

		return this.x == shape.x && this.y == shape.y && this.color.equals(shape.color);
	}

}

class Rectangle extends Shape {

	private int width, height;
	private String color;

	Rectangle(int width, int height) {
		super();
		this.width = width;
		this.height = height;
	}

	private Rectangle(Cloneable prototype) {
		super(prototype);

		Rectangle rec = (Rectangle)prototype;
		this.width = rec.width;
		this.height = rec.height;
	}

	@Override
	public Shape clone() {
		return new Rectangle(this);
	}

	@Override
	public boolean equals(Object another) {
		if (another == null || !(another instanceof Shape)) return false;

		Rectangle rec = (Rectangle)another;

		return super.equals(another) && this.width == rec.width && this.height == rec.height;
	}
}

final class Geometry {

	public static void main(String[] args) {
		Rectangle rec = new Rectangle(10, 30);
		Rectangle rec2 = (Rectangle) rec.clone();

		System.out.printf("Are the same? %s\n", rec.equals(rec2));

	}
}

File: prototype/kotlin/geometry.kt

interface Prototype {

	fun clone(): Prototype
}

abstract class Shape(
	internal var x: Int,
	internal var y: Int,
	internal var color: String
) : Prototype {
	internal constructor() : this(120, 144, "red")

	protected constructor(prototype: Prototype) : this() {
		val shape: Shape = prototype as Shape;
		this.x = shape.x;
		this.y = shape.y;
		this.color = shape.color;
	}

	override abstract fun clone(): Prototype

	override open fun equals(other: Any?): Boolean {
		if (other == null || other !is Shape) return false

		return this.x == other.x && this.y == other.y
			&& this.color.equals(color)
	}
}

class Rectangle : Shape {

	var width: Int
		private set
	var height: Int
		private set

	internal constructor(width: Int, height: Int) : super() {
		this.width = width
		this.height = height
	}

	private constructor(prototype: Prototype) : super(prototype) {
		val rec: Rectangle = prototype as Rectangle
		this.width = rec.width
		this.height = rec.height
	}

	override fun clone(): Prototype = Rectangle(this)

	override fun equals(other: Any?): Boolean {
		if (other == null || other !is Rectangle) return false

		return super.equals(other) && this.width == other.width
			&& this.height == other.height
	}
}

class Circle : Shape {

	var radius: Int
		private set

	internal constructor(radius: Int) : super() {
		this.radius = radius
	}

	private constructor(prototype: Prototype) : super(prototype) {
		val circle: Circle = prototype as Circle
		this.radius = circle.radius
	}

	override fun clone(): Prototype = Circle(this)

	override fun equals(other: Any?): Boolean {
		if (other == null || other !is Circle) return false

		return super.equals(other) && this.radius == other.radius
	}
}

fun main(): Unit {
	val circle: Circle = Circle(8)
	val circleClone: Circle = circle.clone() as Circle

	val rec: Rectangle = Rectangle(15, 18)
	val recClone: Rectangle = rec.clone() as Rectangle

	println("Are the rectangles same? ${rec.equals(recClone)}")
	println("Are the circles same? ${circle.equals(circleClone)}")
}